kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java

```diff
@@ -250,43 +261,61 @@ private boolean checkRotationOrAppend(
       String encodedPartition,
       long now
   ) {
-    if (compatibility.shouldChangeSchema(record, null, currentValueSchema)
-        && recordCount > 0) {
-      // This branch is never true for the first record read by this TopicPartitionWriter
-      log.trace(
-          "Incompatible change of schema detected for record '{}' with encoded partition "
-          + "'{}' and current offset: '{}'",
-          record,
-          encodedPartition,
-          currentOffset
-      );
-      currentSchemas.put(encodedPartition, valueSchema);
-      nextState();
-    } else if (rotateOnTime(encodedPartition, currentTimestamp, now)) {
-      setNextScheduledRotation();
-      nextState();
-    } else {
-      currentEncodedPartition = encodedPartition;
-      SinkRecord projectedRecord = compatibility.project(
-          record,
-          null,
-          currentValueSchema
-      );
-      writeRecord(projectedRecord);
-      buffer.poll();
-      if (rotateOnSize()) {
-        log.info(
-            "Starting commit and rotation for topic partition {} with start offset {}",
-            tp,
-            startOffsets
+    try {
+      if (compatibility.shouldChangeSchema(record, null, currentValueSchema)
+          && recordCount > 0) {
+        // This branch is never true for the first record read by this TopicPartitionWriter
+        log.trace(
+            "Incompatible change of schema detected for record '{}' with encoded partition "
+                + "'{}' and current offset: '{}'",
+            record,
+            encodedPartition,
+            currentOffset
         );
+        currentSchemas.put(encodedPartition, valueSchema);
+        nextState();
+      } else if (rotateOnTime(encodedPartition, currentTimestamp, now)) {
+        setNextScheduledRotation();
         nextState();
-        // Fall through and try to rotate immediately
       } else {
-        return false;
+        currentEncodedPartition = encodedPartition;
+        SinkRecord projectedRecord = compatibility.project(
+            record,
+            null,
+            currentValueSchema
+        );
+        writeRecord(projectedRecord);
+        buffer.poll();
+        if (rotateOnSize()) {
+          log.info(
+              "Starting commit and rotation for topic partition {} with start offset {}",
+              tp,
+              startOffsets
+          );
+          nextState();
+          // Fall through and try to rotate immediately
+        } else {
+          return false;
+        }
+      }
+      return true;
+    } catch (SchemaProjectorException e) {
+      if (reporter != null) {
+        reporter.report(record, e);
+        log.warn("Errant record written to DLQ due to SchemaProjectorException.");
+        buffer.poll();
+        return true;
+      } else {
+        throw new ConnectException(e);
+      }
+    } catch (ConnectException e) {
```

Comment:
It may be useful to report the exceptions due to the keys/headers not being present when expected. These are thrown in [the KeyValueHeaderRecordWriterProvider](https://github.com/confluentinc/kafka-connect-storage-cloud/blob/f42294dae9f94f0cb258f05a3ec7b3b9c664652d/kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/KeyValueHeaderRecordWriterProvider.java#L91)